Split model functionality out into DTOs and passport strategy

rmainwork opened the pull request:
The model is currently being used to fulfill pieces of functionality that would be a better fit elsewhere in the application

In-keeping with the single-responsibility principle:

- SAML strategy should be responsible for re-mapping the response from HarvardKey to the format expected by our application(this PR remaps the data from HK into a new instance of the `User` entity)
- Typing of server response should be done with DTOs, rather than trying to re-use the model for this and pulling it into the client

## Types of changes
<!--
  What types of changes does your code introduce? Put an `x` in all the boxes
  that apply:
-->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality
      to change)

## Checklist:
<!--
  Go over all the following points, and put an `x` in all the boxes that apply.
-->
- [X] I have run `eslint` on the code
- [x] I have added JSDoc for all of my code (where applicable)
- [x] I have added tests to cover my changes.

## Priority:
- [X] Normal <!-- New piece of functionality -->
- [ ] High <!-- Critical bug requiring urgent review -->

## Related Issues:
<!--
  Add the issue number in below that this PR is intended to address.

  Also mention any issues that this PR is related to, and if possible,
  provide more information regarding the relationship to the issues in
  question (i.e does the PR fix the issue, is it designed to fix another bug
  that is similar to the issue etc.)
-->
Fixes #10 

<!--
  Attribution:
  PR Template adapted from: https://github.com/h5bp/html5-boilerplate/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

Commit:
Re-map user data in the passport strategy rather than the model

In-keeping with the single-responsibility principle, the SAML strategy should be responsible for re-mapping the response from HarvardKey to the format expected by our application

Commit:
NestJS Testing

Bring nest testing utils in so that we can bring up a mock IoC container

Commit:
SAML strategy rejects throws unauthorized exception for failed auth

Commit:
SAML strategy returns dummy user when in dev mode

Commit:
Remove HK profile interface from user class

Commit:
Imports

Replace relative path to test data with alias and change ordering to fix eslint

Commit:
Client-side DTO

Replace client dependency on user model with dependency on client-side DTO. A separate set of DTOs are used on the client and server to preserve process boundaries and to allow things like validation to be done using the server-side DTOs without impacting the client

Commit:
Server-side DTO

Implement server-side DTO to describe the format of a user object as it is returned to the client

Commit:
Remove lock file

Package lock file shouldn't be in the boilerplate repo for compatibility reasons

Commit:
Config service tests

Add tests to cover behaviour in configuration service and increase the global app coverage level to >=80%

Commit:
eslint

jonseitz commented on src/client/api/users.ts:
```diff
@@ -1,5 +1,5 @@
 import request, { AxiosPromise } from 'axios';
-import { User } from '../../server/models/user.class';
+import { UserResponse as User } from 'client/dto/users/userResponse.dto';
```
I think it would be better practice to keep the name as `UserResponse` rather than aliasing it to `User` -- Not super relevant now, but I could see this getting confusing if we were importing multiple `dto`s.

rmainwork replied:
I agree. I was just trying to keep the diff sane. Updated in 484e5ac

jonseitz commented on src/client/components/App.tsx:
```diff
@@ -6,14 +6,14 @@ import React, {
   UserContext,
 } from 'client/context';
 import { getCurrentUser } from 'client/api';
+import { UserResponse as User } from 'client/dto/users/userResponse.dto';
```
as above, suggest keeping the import name.

rmainwork replied:
Updated in 484e5ac

jonseitz commented on src/client/components/__tests__/App.test.tsx:
```diff
@@ -3,7 +3,7 @@ import { strictEqual } from 'assert';
 import { stub, SinonStub } from 'sinon';
 import { AxiosResponse } from 'axios';
-import { User } from 'server/models';
+import { UserResponse as User } from 'client/dto/users/userResponse.dto';
```
as above, suggest keeping the import name.

rmainwork replied:
Updated in 484e5ac

jonseitz commented on src/client/context/UserContext.ts:
```diff
@@ -1,5 +1,5 @@
 import { createContext, Context } from 'react';
-import { User } from '../../server/models/user.class';
+import { UserResponse as User } from 'client/dto/users/userResponse.dto';
```
as above, suggest keeping the import name.

rmainwork replied:
Updated in 484e5ac

jonseitz commented on src/server/services/__tests__/config.service.test.ts:
```diff
@@ -0,0 +1,50 @@
+
+describe('Configuration Service', function () {
+  it('reports if the app is in production', async function () {
+    process.env.NODE_ENV = 'production';
```
Potential issue here: Our tests normally run in `NODE_ENV === 'testing'`, which helps for things like not sending email or connecting to different databases. Changing the `NODE_ENV` could then affect any tests that come after this.

I think the best way to resolve this is to rewrite the config service constructor so it reads the environment from a passed object, instead of directly from `process`. That will also make it easier if we need to test other values in the config service in the future.

rmainwork replied:
>Potential issue here: Our tests normally run in NODE_ENV === 'testing', which helps for things like not sending email or connecting to different databases. Changing the NODE_ENV could then affect any tests that come after this.


That's a good point, I'd forgotten about that. 


> I think the best way to resolve this is to rewrite the config service constructor so it reads the environment from a passed object, instead of directly from process. That will also make it easier if we need to test other values in the config service in the future.

Yeah, I wonder if we should tweak it a little so that it basically just parses the `.env` file straight in there and we can just read and map arbitrary values into it rather than having a defined schema for `DB_HOST` etc. What are your thoughts?

jonseitz replied:
Just to recap our offline conversation for the official record: The `.env` file is being parsed by docker-compose, not the application code. That way we can automate building the image on one system (i.e. Travis) and let the deployment details be handled by the production system (i.e. Fargate). So the easiest way for us to pass arbitrary values into the `ConfigService` is for the constructor to accept an object, which in production will just be the value `process.env` and in testing could be any set of values we want to test against.

rmainwork replied:
Addressed in 2ce2cd5dfd7efdcc7edc407452cec9356cd90058

rmainwork commented:
Coming back to this, I'm wondering if it makes sense to move the `HarvardKeyProfile` in with the SAML strategy and export it from there? Since it's not really being outside outside of this context..

jonseitz commented on src/server/dto/users/userResponse.dto.ts:
```diff
@@ -0,0 +1,25 @@
+ * users etc.)
+ */
+
+export abstract class UserResponse {
```
I'm a little concerned with having the same dto defined in two different files -- that seems like a lot of potential for error if we want to change things in the future. Example: say we decide to use `surname` instead of `lastName`, after we've already created multiple `User*` dtos. That's a lot to update for one small change...

What if we define our shared `dto`'s in the common directory, then we could re-export them (or `extend` them if we need to make changes) from the `client`/`server` directories?

rmainwork replied:
I know what you mean and I kind of agree. At the same time, I would be wary of doing that for a couple of reasons...

1. There are things we're going to end up with in our server-side DTOs(for example validation with [class-validator](https://github.com/typestack/class-validator)) that don't necessarily make sense on the client side and possibly vice versa...
1. If our API changes field names etc. we would have to do refactoring anyway of UI components, database entities etc. so this would just shift the boundary of the problem...

rmainwork replied:
Fixed in 2a86c48d4a7502dbdf281a9f7c621ef773bcddc3

rmainwork replied:
...and 2fd2139a17b7f94af62ba72971026473e93c5b03

jonseitz reviewed (changes requested):
Just a couple changes here, mostly things that _might_ help keep us sane in the future...

Commit:
Don't Alias UserResponse DTO

For sanity, it makes more sense to refer to the plain use DTO as
`UserResponse` and not `User`(to avoid confusion with the `User` entity
etc.)

Commit:
Move client and server user DTO into common folder

Commit:
Fix paths to shared user DTO

Commit:
Functional config service

Refactor config service to allow it to wrap arbitrary key value objects and use them as configuration(instead of ingesting `process.env` directly and thus having a hard dependency on it)

Commit:
Add test for mongooseOptions getter

jonseitz commented:
Approving this PR despite failure on CodeCov. As best we can tell, it's detecting a coverage "miss" in the constructor of the Config service. This may be because the constructor signature includes a default value of `{}`, but Robert and I both agree this falls into the category of a language feature and doesn't need to be tested. 

In a future PR, we'll look at adjusting the failure threshold for codecov to allow for more wiggle room in cases like this. While it's good in principle to not have coverage levels change between PRs, it doesn't make sense to have a small blip like this derail a PR by requiring changes outside the scope of the original issue.
